### PR TITLE
Allow server to redirect clients

### PR DESCRIPTION
**Pull request by ChillerDragon**
This is a preperation for #6754 (redirect players on map vote) And an implementation of #5662 (redirect players)

Reopen of #6757

## Checklist

- [x] Tested the change ingame
- [ ] Provided screenshots if it is a visual change
- [ ] Tested in combination with possibly related configuration options
- [ ] Written a unit test (especially base/) or added coverage to integration test
- [ ] Considered possible null pointers and out of bounds array indexing
- [x] Changed no physics that affect existing maps
- [ ] Tested the change with [ASan+UBSan or valgrind's memcheck](https://github.com/ddnet/ddnet/#using-addresssanitizer--undefinedbehavioursanitizer-or-valgrinds-memcheck) (optional)

### CODE REVIEW

**Review by heinrich5991 — CHANGES_REQUESTED** [review]
Should have a server implementation.

**Comment by ChillerDragon** [comment]
https://github.com/ddnet/ddnet/pull/6757#issuecomment-1648665077
> > Okay easy fix. Delete more code and make this a client only pullrequest stuck_out_tongue_winking_eye it's the more important part anyways to get it out to clients. Server Code can be changed easily
> 
> This doesn't help with the evil lazy people. :p Do you want an implementation idea for ignoring further messages/dropping old clients?

Its mostly that I disagree that a redirect should be enforced. Also in my opinion it depends on why the server implementation redirects. As per implementation idea I would just do a ``ExecuteLine("kick %d redirected to %s", ClientID, pAddress);``

**Comment by ChillerDragon** [comment]
I ignored packets after the connection is dropped server side as @heinrich5991 suggested. And I very much dislike the outcome. The client keeps sending input packets and thinks its connected. There is no connection problems message shown. The server keeps sending snapshots. The connection does not time out. Here a tshark output of the client ignoring a redirect packet.

```
    1   0.000000 192.168.178.78 42640 192.168.178.78 8303 TW 67 sys.input
    2   0.058285 192.168.178.78 42640 192.168.178.78 8303 TW 67 sys.input
    3   0.099947 192.168.178.78 42640 192.168.178.78 8303 TW 67 sys.input
    4   0.158282 192.168.178.78 42640 192.168.178.78 8303 TW 67 sys.input
    5   0.216728 192.168.178.78 42640 192.168.178.78 8303 TW 67 sys.input
    6   0.258399 192.168.178.78 42640 192.168.178.78 8303 TW 67 sys.input
    7   0.316746 192.168.178.78 42640 192.168.178.78 8303 TW 67 sys.input
    8   0.374967 192.168.178.78 42640 192.168.178.78 8303 TW 67 sys.input
    9   0.416618 192.168.178.78 42640 192.168.178.78 8303 TW 67 sys.input
   10   0.451492 192.168.178.78 8303 192.168.178.78 42640 TW 386 sys.snap_single
   11   0.483287 192.168.178.78 42640 192.168.178.78 8303 TW 67 sys.input
   12   0.524950 192.168.178.78 42640 192.168.178.78 8303 TW 67 sys.input
   13   0.583281 192.168.178.78 42640 192.168.178.78 8303 TW 68 sys.input
   14   0.624949 192.168.178.78 42640 192.168.178.78 8303 TW 68 sys.input
   15   0.683282 192.168.178.78 42640 192.168.178.78 8303 TW 67 sys.input
   16   0.741620 192.168.178.78 42640 192.168.178.78 8303 TW 68 sys.input
   17   0.783293 192.168.178.78 42640 192.168.178.78 8303 TW 68 sys.input
   18   0.841620 192.168.178.78 42640 192.168.178.78 8303 TW 67 sys.input
   19   0.899950 192.168.178.78 42640 192.168.178.78 8303 TW 68 sys.input
```

Just seems like a debug nightmare if for whatever reason a client ends up in that state. The ignoring packets instead of a full on connection kill was because I think the client should be able to receive resend redirect packets if one gets lost. Instead of being kicked without knowing what to do. But not sure if it is worth this mess.

**Comment by heinrich5991** [comment]
The client shouldn't end up in that state anyway. This is similar to what happens if, say, the client misses a message while the connection is being established. The client will not recover from that.

If you wanted to, you could go lower-level and make sure the server completely ignores the the client connection except for resending.

**Comment by ChillerDragon** [comment]
> The client shouldn't end up in that state anyway. This is similar to what happens if, say, the client misses a message while the connection is being established. The client will not recover from that.
> 
> 
> 
> If you wanted to, you could go lower-level and make sure the server completely ignores the the client connection except for resending.

I sometimes get stuck in the connection screen while already being able to receive messages. Yeah the client will not recover but idle in this weird state.

Going lower level sounds like adding more complexity to the network code. That seems like a bad thing. And I still did not understand the advantage of dropping the connection in the first place.

**Comment by heinrich5991** [comment]
> And I still did not understand the advantage of dropping the connection in the first place.

The advantage is enforcing the protocol. Otherwise someone funny will come along the way and create a client to ignore redirect messages.

**Comment by heinrich5991** [comment]
4

> I sometimes get stuck in the connection screen while already being able to receive messages. Yeah the client will not recover but idle in this weird state.

If that happens, a pcap and a bug report would be appreciated.

**Comment by ChillerDragon** [comment]
> > I sometimes get stuck in the connection screen while already being able to receive messages. Yeah the client will not recover but idle in this weird state.
> 
> 
> 
> If that happens, a pcap and a bug report would be appreciated.

It's not like I can reproduce it. But you have never seen that?

**Comment by heinrich5991** [comment]
> But you have never seen that?

Not sure I've seen this outside of the "DoS-protected" servers. A pcap while the client is already in the state would probably also be interesting.

**Comment by ChillerDragon** [comment]
> > But you have never seen that?
> 
> Not sure I've seen this outside of the "DoS-protected" servers. A pcap while the client is already in the state would probably also be interesting.

offtopic moment. But yeah now you made me doubt that I saw it outside of ddos protection too. I will giving the network refactor a go but im slowly leaning towards kicking clients without resend support to keep it clean.

**Comment by ChillerDragon** [comment]
Now looks like this

```
  121   4.562827 21247 8303 TW 98 game.cl_say, sys.input (me sending a chat command that triggers redirect)
  122   4.563006 8303 21247 TW 90 sys.4efe406a-7774-33f1-bfde-1806ff6d1528 (redirect msg which my client ignores)
  123   4.623241 21247 8303 TW 84 sys.input
  124   4.683713 21247 8303 TW 84 sys.input
  125   4.744112 21247 8303 TW 84 sys.input
  126   4.802467 21247 8303 TW 84 sys.input
  127   4.844125 21247 8303 TW 84 sys.input
  128   4.902451 21247 8303 TW 84 sys.input
  129   4.944111 21247 8303 TW 84 sys.input
  130   5.004505 21247 8303 TW 84 sys.input
  131   5.050391 21247 8303 TW 84 sys.input
  132   5.078691 8303 21247 TW 152 game.sv_chat, game.sv_chat (unrelated)
  133   5.108627 21247 8303 TW 84 sys.input
  134   5.150300 21247 8303 TW 84 sys.input
  135   5.208597 21247 8303 TW 84 sys.input
  136   5.250291 21247 8303 TW 84 sys.input
  137   5.308683 21247 8303 TW 84 sys.input
  138   5.350301 21247 8303 TW 84 sys.input
  139   5.408732 21247 8303 TW 84 sys.input
  140   5.450410 21247 8303 TW 84 sys.input
  141   5.508733 21247 8303 TW 84 sys.input
  142   6.078811 8303 21247 TW 70 ctrl.keep_alive
  143   6.517077 21247 8303 TW 70 ctrl.keep_alive
  144   7.078833 8303 21247 TW 70 ctrl.keep_alive
  145   7.525441 21247 8303 TW 70 ctrl.keep_alive
  146   8.098826 8303 21247 TW 70 ctrl.keep_alive
  147   8.525402 21247 8303 TW 70 ctrl.keep_alive
  148   9.098815 8303 21247 TW 70 ctrl.keep_alive
  149   9.525548 21247 8303 TW 70 ctrl.keep_alive
  150  10.118804 8303 21247 TW 70 ctrl.keep_alive
  151  10.533732 21247 8303 TW 70 ctrl.keep_alive
  152  11.138804 8303 21247 TW 70 ctrl.keep_alive
  153  11.533904 21247 8303 TW 70 ctrl.keep_alive
  154  12.158809 8303 21247 TW 70 ctrl.keep_alive
  155  12.542044 21247 8303 TW 70 ctrl.keep_alive
  156  13.158821 8303 21247 TW 70 ctrl.keep_alive
  157  13.550396 21247 8303 TW 70 ctrl.keep_alive
  158  14.178806 8303 21247 TW 70 ctrl.keep_alive
  159  14.550543 21247 8303 TW 70 ctrl.keep_alive
  160  14.578858 8303 21247 TW 81 ctrl.disconnect (server triggered timeout after 10 seconds)
```

Wouldn't call it perfect but no idea how to make it cleaner.

File: src/engine/server/server.cpp [hunk 1]
```diff
@@ -520,6 +522,35 @@ void CServer::Ban(int ClientID, int Seconds, const char *pReason)
...
+	bool SupportsRedirect = GetClientVersion(ClientID) >= VERSION_DDNET_REDIRECT;
+	if(Verbose)
+	{
+		str_format(aBuf, sizeof(aBuf), "redirecting '%s' to port %d supported=%d", ClientName(ClientID), Port, SupportsRedirect);
+		Console()->Print(IConsole::OUTPUT_LEVEL_STANDARD, "redirect", aBuf);
+	}
+
+	if(!SupportsRedirect)
+	{
+		str_format(aBuf, sizeof(aBuf), "redirected to port %d", Port);
```
Comment (heinrich5991) [inline src/engine/server/server.cpp:540]: I think this should be rephrased to "Redirect unsupported: please reconnect" if the port is the current server's port or "Redirect unsupported: please connect to port %d" if the port is not the current server's port.

File: src/engine/server/server.cpp [hunk 2]
```diff
@@ -2435,6 +2470,10 @@ void CServer::PumpNetwork(bool PacketWaiting)
 				{
 					GameFlags |= MSGFLAG_VITAL;
 				}
+				if(m_aClients[Packet.m_ClientID].m_Redirected)
```
Comment (heinrich5991) [inline src/engine/server/server.cpp:2473]: This should go above the `int GameFlags = 0` line.

**Review by heinrich5991 — APPROVED** [review]
Looks good to me now. Could you address the two final nits and this should be good to go. :)

**Comment by ChillerDragon** [comment]
Also removed the redundant ``m_Redirected`` variable since ``m_State = CClient::STATE_REDIRECTED`` already stores that info.

**Comment by heinrich5991** [comment]
Thanks!

**Comment by ChillerDragon** [comment]
fyi @def- I predicted the next ddnet release to be 17200

https://github.com/ddnet/ddnet/pull/6914/files#diff-feeef7dae2f4f406a289d06332027d2170618eab5efe8f758e983d5dc2ff9e0fR127

**Pull request merged**